File: js/moq/src/ietf/subscribe_namespace.ts
import type * as Path from "../path.ts";
import type { Reader, Writer } from "../stream.ts";
import * as Message from "./message.ts";
import * as Namespace from "./namespace.ts";

// In draft-14, SUBSCRIBE_ANNOUNCES is renamed to SUBSCRIBE_NAMESPACE
export class SubscribeNamespace {
	static id = 0x11;

	namespace: Path.Valid;
	requestId: number;

	constructor(namespace: Path.Valid, requestId: number) {
		this.namespace = namespace;
		this.requestId = requestId;
	}

	async #encode(w: Writer): Promise<void> {
		await Namespace.encode(w, this.namespace);
		await w.u53(this.requestId);
		await w.u8(0); // no parameters
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<SubscribeNamespace> {
		return Message.decode(r, SubscribeNamespace.#decode);
	}

	static async #decode(r: Reader): Promise<SubscribeNamespace> {
		const namespace = await Namespace.decode(r);
		const requestId = await r.u53();

		const numParams = await r.u8();
		if (numParams !== 0) {
			throw new Error(`SUBSCRIBE_NAMESPACE: parameters not supported: ${numParams}`);
		}

		return new SubscribeNamespace(namespace, requestId);
	}
}

export class SubscribeNamespaceOk {
	static id = 0x12;

	requestId: number;

	constructor(requestId: number) {
		this.requestId = requestId;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u53(this.requestId);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<SubscribeNamespaceOk> {
		return Message.decode(r, SubscribeNamespaceOk.#decode);
	}

	static async #decode(r: Reader): Promise<SubscribeNamespaceOk> {
		const requestId = await r.u53();
		return new SubscribeNamespaceOk(requestId);
	}
}

export class SubscribeNamespaceError {
	static id = 0x13;

	requestId: number;
	errorCode: number;
	reasonPhrase: string;

	constructor(requestId: number, errorCode: number, reasonPhrase: string) {
		this.requestId = requestId;
		this.errorCode = errorCode;
		this.reasonPhrase = reasonPhrase;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u53(this.requestId);
		await w.u62(BigInt(this.errorCode));
		await w.string(this.reasonPhrase);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<SubscribeNamespaceError> {
		return Message.decode(r, SubscribeNamespaceError.#decode);
	}

	static async #decode(r: Reader): Promise<SubscribeNamespaceError> {
		const requestId = await r.u53();
		const errorCode = Number(await r.u62());
		const reasonPhrase = await r.string();

		return new SubscribeNamespaceError(requestId, errorCode, reasonPhrase);
	}
}

export class UnsubscribeNamespace {
	static id = 0x14;

	requestId: number;

	constructor(requestId: number) {
		this.requestId = requestId;
	}

	async #encode(w: Writer): Promise<void> {
		await w.u53(this.requestId);
	}

	async encode(w: Writer): Promise<void> {
		return Message.encode(w, this.#encode.bind(this));
	}

	static async decode(r: Reader): Promise<UnsubscribeNamespace> {
		return Message.decode(r, UnsubscribeNamespace.#decode);
	}

	static async #decode(r: Reader): Promise<UnsubscribeNamespace> {
		const requestId = await r.u53();
		return new UnsubscribeNamespace(requestId);
	}
}

// Backward compatibility aliases
export const SubscribeAnnounces = SubscribeNamespace;
export const SubscribeAnnouncesOk = SubscribeNamespaceOk;
export const SubscribeAnnouncesError = SubscribeNamespaceError;
export const UnsubscribeAnnounces = UnsubscribeNamespace;

File: rs/moq/src/ietf/subscribe_namespace.rs
//! IETF moq-transport-14 subscribe namespace messages

use std::borrow::Cow;

use crate::{coding::*, ietf::Message, Path};

use super::namespace::{decode_namespace, encode_namespace};

/// SubscribeNamespace message (0x11)
#[derive(Clone, Debug)]
pub struct SubscribeNamespace<'a> {
	pub namespace: Path<'a>,
	pub request_id: u64,
}

impl<'a> Message for SubscribeNamespace<'a> {
	const ID: u64 = 0x11;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		encode_namespace(w, &self.namespace);
		self.request_id.encode(w);
		0u8.encode(w); // no parameters
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let namespace = decode_namespace(r)?;
		let request_id = u64::decode(r)?;

		// Ignore parameters, who cares.
		let _params = Parameters::decode(r)?;

		Ok(Self { namespace, request_id })
	}
}

/// SubscribeNamespaceOk message (0x12)
#[derive(Clone, Debug)]
pub struct SubscribeNamespaceOk {
	pub request_id: u64,
}

impl Message for SubscribeNamespaceOk {
	const ID: u64 = 0x12;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.request_id.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let request_id = u64::decode(r)?;
		Ok(Self { request_id })
	}
}
/// SubscribeNamespaceError message (0x13)
#[derive(Clone, Debug)]
pub struct SubscribeNamespaceError<'a> {
	pub request_id: u64,
	pub error_code: u64,
	pub reason_phrase: Cow<'a, str>,
}

impl<'a> Message for SubscribeNamespaceError<'a> {
	const ID: u64 = 0x13;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.request_id.encode(w);
		self.error_code.encode(w);
		self.reason_phrase.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let request_id = u64::decode(r)?;
		let error_code = u64::decode(r)?;
		let reason_phrase = Cow::<str>::decode(r)?;

		Ok(Self {
			request_id,
			error_code,
			reason_phrase,
		})
	}
}

/// UnsubscribeNamespace message (0x14)
#[derive(Clone, Debug)]
pub struct UnsubscribeNamespace {
	pub request_id: u64,
}

impl Message for UnsubscribeNamespace {
	const ID: u64 = 0x14;

	fn encode<W: bytes::BufMut>(&self, w: &mut W) {
		self.request_id.encode(w);
	}

	fn decode<R: bytes::Buf>(r: &mut R) -> Result<Self, DecodeError> {
		let request_id = u64::decode(r)?;
		Ok(Self { request_id })
	}
}
